procServUtils autocomplete and new commands

darcato opened the pull request:
Hello, 
this PR adds some functionalities to the procServUtils scripts:

1. Bash autocomplete through `argcomplete`
1. New commands to restart, rename or open logs of the processes
1. Print `status` as a table with color support, to be easier to read
1. Fix some unhandled exceptions
1. Documentation for the commands

To enable these functionalities three python dependencies are needed: `argcomplete`, `tabulate` and `termcolor`. Unfortunately the current installation process does not support installing dependencies so one has to install them manually through pip.

Do you think the installation process can be modified to use pip to automatically handle dependencies?

All the changes here are compatible with existing configuration files and do not break compatibility (hopefully).

darcato committed:
Autocomplete with argcomplete

darcato committed:
Autocomplete instances names

darcato committed:
Get user configuration during autocomplete

darcato committed:
Adding logs and restart commands to manage-procs

giosava94 committed:
Reset a failed procServ instance

giosava94 committed:
Insert reset-failed in remove func

darcato committed:
Merge pull request #1 from giosava94/dev

Reset-failed when removing procs

darcato committed:
Improving unit removal, fixing some exceptions

darcato committed:
Pretty table on manage-procs status

darcato committed:
Merge branch 'dev'

darcato committed:
Renameprocs

darcato committed:
Addproc with new addconf utility

darcato committed:
Fix rename on unknown process

darcato committed:
Check existence of requested proc

darcato committed:
Comment with dependencies

darcato committed:
Merge branch 'dev'

darcato committed:
Merge branch 'master' of https://github.com/ralphlange/procServ

darcato committed:
Double import

darcato committed:
Adding procServUtils docs

darcato committed:
Link new readmi

darcato committed:
Typos

darcato committed:
Fix code review

darcato committed:
Fix travis

darcato committed:
Slightly different conf file spacing

darcato committed:
Whitespaces

darcato committed:
Fix manage-procs rename docs

mdavidsaver commented:
My first benchmark for dependencies is presence in the Debian package repository.  All three have been present since at least Debian 8 (10 is stable atm).  [argcomplete](https://packages.debian.org/python3-argcomplete) [tabulate](https://packages.debian.org/python3-tabulate) [termcolor](https://packages.debian.org/python3-termcolor)

[argcomplete](https://pypi.org/project/argcomplete/) and [tabulate](https://pypi.org/project/tabulate/) appear at a glance to be healthy projects with recent releases and links to VCS repos.

termcolor appears to be defunct, with [no release since 2011](https://pypi.org/project/termcolor/#history), and no authoritative VCS repository listed.

mdavidsaver commented:
Personally, I would make this three optional dependencies with some `try: ... import`.  A used, termcolor and argcomplete can be trivially stubbed out, with tabulate being only slightly more involved.

mdavidsaver commented on procServUtils/manage.py:
```diff
@@ -207,11 +222,18 @@ def delproc(conf, args):
 
     _log.info("Disabling service procserv-%s.service", args.name)
-    SP.check_call([systemctl,
+    SP.call([systemctl,
```
Why change to the un-checked call?  Have you seen this step fail?  As a general rule I don't like the idea of ignoring failures.

darcato replied:
I remember seeing this fail, but I cannot reproduce the problem now (it was months ago), so for the moment I will revert back to the ``check_call`` method.

mdavidsaver commented on procServUtils/manage.py:
```diff
@@ -226,6 +248,46 @@ def delproc(conf, args):
+    _log.info('Trigger systemd reload')
+    SP.check_call([systemctl,
+                   argusersys,
+                   'daemon-reload'], shell=False)
```
I'm not sure about the ordering here, but I think the `daemon-reload` should come first.  As I understand it, a reload is needed after manually editing any systemd config files, eg. before any new units are recognised.

I'm also wondering it an explicit `disable` is needed for the original service name?  Have you verified that the symlinks are being cleaned up?  (eg. `ls -l /etc/systemd/system/multi-user.target.wants/`)

mark0n replied:
I agree with @mdavidsaver, the order should be:
1. stop and disable old unit (calling disable is required if we don't want to leave any mess behind)
2. modify unit files
3. daemon-reload
4. enable and start new unit

darcato replied:
Ok, good catch! I copied this logic from the ``addproc`` function, so I will fix the order also on that function.

mdavidsaver commented on setup.py:
```diff
@@ -29,4 +29,5 @@ def run(self):
         ('lib/systemd/user-generators', ['systemd-procserv-generator-user']),
     ],
     cmdclass = { 'install_data': custom_install_data },
+    #install_requires=['argcomplete', 'tabulate', 'termcolor'],
```
If these are to be unconditional dependencies, then `install_requires` needs to be uncommentated.

darcato replied:
The problem here is that the project uses ``distutils`` to install the python scripts, which does not support the ``install_requires`` directive. Maybe it's possible to use ``pip``, but I'm not sure how to integrate it with the makefiles.

For the moment I will follow your suggestion to make this packages optional using the ``try: ... import`` mechanism and defining some fallback behaviors.

mdavidsaver replied:
Ya...  I was a late and reluctant convert to using setuptools.  I had hoped that distutils would be cleaned up to include the useful parts.  However, the opposite has happened.  With so much external tooling with invasive hooks into distutils, it doesn't seem like it can reasonable be changed.  At least setuptools availability is ubiquitous now.

Although I don't think this module will ever be usable in eg. a virtualenv, switching to setuptools is probably reasonable.  Doing this may necessitate changes to the `custom_install_data` though.

darcato committed:
Conditional usage of python packages termcolor, tabulate and argcomplete

darcato commented:
Hello @mdavidsaver,
thanks for your comments. I will try to address all of them in the following days.

darcato committed:
Fix systemctl call order

darcato committed:
Revert check call on systemctl disable

darcato committed:
Updating procServUtils doc

darcato commented on procServUtils/README.md:
````diff
@@ -0,0 +1,99 @@
+```
+Activate global arcomplete completion:
+```bash
+sudo activate-global-python-argcomplete 
````
Ok, I will list it as an optional step!

darcato committed:
Whitespaces

darcato committed:
Fix devnull for python2

darcato committed:
Add test for rename command

darcato commented:
I made the three packages optional, adding some fallback functions that emulate almost the same behavior as before this PR.
Then I fixed the ``systemctl`` calls order and reverted the ``check_call`` on ``systemctl disable``. I also added a test function for the ``rename`` command and updated the readme.

I redirected the standard error of the ``systemctl reset-failed`` command to ``/dev/null`` because this fails when there is nothing to reset (which is the most common situation).

I didn't remove the ``install_requires`` comment, as this will be useful when the project will move to ``setuptools``. For the moment I did not implement this change.

darcato committed:
Ask before stopping service to rename

darcato committed:
Fix test errors

darcato commented:
Lastly, since the ``rename`` command causes the service to stop, I added a dialog for the user to confirm the operation, and I added a notice to the readme.

ralphlange commented:
@mdavidsaver: Would you think the added features are important enough to justify rebasing this to the modernized python setup?
(Remember that we don't use procServ nor the utilities, so it's really hard for me to tell.)

mdavidsaver commented:
I think the installation of procServUtils needs to be rethought.  eg. to install outside of the default PYTHONPATH as a private module as `/usr/lib/procServ/procServUtils`.

It think it would be possible to have autotools do this, which would avoid any resort to setuptools.
